### PR TITLE
Sprint 2

### PR DESCRIPTION
**Pull request by simonstadlinger**

### CODE REVIEW

File: src/lib/storage/table.cpp [hunk 1]
```diff
@@ -81,23 +83,71 @@ const std::string& Table::column_name(const ColumnID column_id) const { return _
...
+
+  return compressed_chunk;
+}
+
+void Table::_compress_column(Chunk& uncompressed_chunk, std::shared_ptr<Chunk>& compressed_chunk,
+                             ColumnID col_id) const {
+  const auto& column_segment = uncompressed_chunk.get_segment(col_id);
+
+  resolve_data_type(column_type(col_id), [&](const auto data_type_t) {
+    using ColumnDataType = typename decltype(data_type_t)::type;
```
Comment (Jasperhino) [inline src/lib/storage/table.cpp:147]: Maybe use a different name than `data_type_t` because the `_t` indicates that it holds a native type.

File: src/test/storage/table_test.cpp [hunk 2]
```diff
@@ -89,4 +89,21 @@ TEST_F(StorageTableTest, PrivateAddSegment) {
...
+  EXPECT_THROW(compressed_chunk.append({3, "Invalid Append"}), std::runtime_error);
+}
+
+TEST_F(StorageTableTest, CompressOnlyFullChunks) {
+  t.append({1, "Insufficient Row"});
+
+  EXPECT_THROW(t.compress_chunk(ChunkID{0}), std::exception);
+}
+
 }  // namespace opossum
```
Comment (Jasperhino) [inline src/test/storage/table_test.cpp:109]: The method `const Chunk& Table::get_chunk(ChunkID chunk_id) const` is not covered in your tests.

**Review by Jasperhino — CHANGES_REQUESTED** [review]
Overall nice implementation 👌 We even managed to find some bugs in our mutex implementation thanks to you 🎉

File: src/lib/storage/dictionary_segment.hpp [hunk 3]
```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
...
+    auto dictionaryVecMem = _dictionary->size() * sizeof(T);
+    return static_cast<size_t>(attributeVecMem + dictionaryVecMem);
+  }
 
  protected:
   std::shared_ptr<std::vector<T>> _dictionary;
   std::shared_ptr<BaseAttributeVector> _attribute_vector;
+
+  void _build_compressed_dictionary(const std::vector<T>& values) {
+    auto raw_dictionary = std::move(values);
```
[the remark this answers is not in the document]
Comment (simonstadlinger) [reply src/lib/storage/dictionary_segment.hpp:112]: We think the ::move is legit, because after that you should not use the values anymore. The value segment is torn down and a dictionary segment built up.

File: src/lib/storage/dictionary_segment.hpp [hunk 4]
```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
...
+    raw_dictionary.erase(std::unique(raw_dictionary.begin(), raw_dictionary.end()), raw_dictionary.end());
+
+    _dictionary = std::make_shared<std::vector<T>>(raw_dictionary);
+
+    _build_attribute_vector(raw_dictionary, raw_values);
+  }
+
+  void _build_attribute_vector(std::vector<T>& raw_dictionary, std::vector<T>& all_values) {
+    std::vector<uint32_t> raw_vector{};
+    raw_vector.reserve(all_values.size());
```
Comment (simonstadlinger) [inline src/lib/storage/dictionary_segment.hpp:125]:
```suggestion
```

File: src/lib/storage/dictionary_segment.hpp [hunk 5]
```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
...
+    } else if (required_width <= 16) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint16_t>>(size);
+    } else {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint32_t>>(size);
+    }
+
+    for (size_t all_values_index = 0; all_values_index < size; all_values_index++) {
+      const auto value = type_cast<T>(all_values.at(all_values_index));
+      auto dictionary_iterator = std::lower_bound(raw_dictionary.begin(), raw_dictionary.end(), value);
+      uint32_t dictionary_index = dictionary_iterator - raw_dictionary.begin();
```
Comment (simonstadlinger) [inline src/lib/storage/dictionary_segment.hpp:139]:
```suggestion
      uint32_t dictionary_index = std::distance(dictionary_iterator, raw_dictionary.begin());
```

File: src/lib/storage/table.cpp (hunk 1 above)
Comment (simonstadlinger) [inline src/lib/storage/table.cpp:147]:
```suggestion
  resolve_data_type(column_type(col_id), [&](const auto col_type_string) {
    using ColumnDataType = typename decltype(col_type_string)::type;
```

File: src/lib/storage/dictionary_segment.hpp [hunk 6]
```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
...
+  ValueID lower_bound(T value) const {
+    auto dictionary_size = _dictionary->size();
+    for (ValueID dictionary_index = ValueID{0}; dictionary_index < dictionary_size; ++dictionary_index) {
+      if (value_by_value_id(dictionary_index) >= value) {
+        return dictionary_index;
+      }
+    }
+
+    return INVALID_VALUE_ID;
+  }
```
Comment (Bensk1) [inline src/lib/storage/dictionary_segment.hpp:74]: The dictionary is sorted, we expect a binary search here.

File: src/lib/storage/dictionary_segment.hpp [hunk 7]
```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
...
 
   // same as lower_bound(T), but accepts an AllTypeVariant
-  ValueID lower_bound(const AllTypeVariant& value) const;
+  ValueID lower_bound(const AllTypeVariant& value) const { return lower_bound(static_cast<T>(value)); }
```
Comment (Bensk1) [inline src/lib/storage/dictionary_segment.hpp:77]: `type_cast` instead of `static_cast`

File: src/lib/storage/dictionary_segment.hpp [hunk 8]
```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
...
 
   // return the number of entries
-  ChunkOffset size() const override;
+  ChunkOffset size() const { return _attribute_vector->size(); }
 
   // returns the calculated memory usage
-  size_t estimate_memory_usage() const final;
+  size_t estimate_memory_usage() const final {
+    auto attributeVecMem = _attribute_vector->width() * _attribute_vector->size();
+    auto dictionaryVecMem = _dictionary->size() * sizeof(T);
```
Comment (Bensk1) [inline src/lib/storage/dictionary_segment.hpp:104]: `attribute_vec_mem` instead of `attributeVecMem`

File: src/lib/storage/dictionary_segment.hpp [hunk 9]
```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
...
   // SEMINAR INFORMATION: Since most of these methods depend on the template parameter, you will have to implement
   // the DictionarySegment in this file. Replace the method signatures with actual implementations.
 
+  // return the value represented by a given ValueID
+  const T& value_by_value_id(ValueID value_id) const { return _dictionary->at(value_id); }
+
   // return the value at a certain position. If you want to write efficient operators, back off!
-  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override;
+  AllTypeVariant operator[](const ChunkOffset chunk_offset) const {
+    return static_cast<AllTypeVariant>(get(chunk_offset));
```
Comment (mweisgut) [inline src/lib/storage/dictionary_segment.hpp:46]: To create an `AllTypeVariant` object that contains a passed value, using the constructor of `AllTypeVariant` is the preferred way. In addition, why did you remove the override?

File: src/lib/storage/dictionary_segment.hpp [hunk 10]
```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
...
  protected:
   std::shared_ptr<std::vector<T>> _dictionary;
   std::shared_ptr<BaseAttributeVector> _attribute_vector;
+
+  void _build_compressed_dictionary(const std::vector<T>& values) {
+    auto raw_dictionary = std::move(values);
+    std::vector<T> raw_values = raw_dictionary;
+
+    std::sort(raw_dictionary.begin(), raw_dictionary.end());
+    raw_dictionary.erase(std::unique(raw_dictionary.begin(), raw_dictionary.end()), raw_dictionary.end());
```
[the remark this answers is not in the document]
Comment (Bensk1) [reply src/lib/storage/dictionary_segment.hpp]: I don't understand this argument.

File: src/lib/storage/dictionary_segment.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bensk1) [reply src/lib/storage/dictionary_segment.hpp]: We agree with @Jasperhino. While the value segment is not used after the dictionary compression has finished, there might be ongoing accesses while the **segment** is compressed. As soon, as L112 is executed these accesses might operate on moved data!

File: src/lib/storage/dictionary_segment.hpp [hunk 11]
```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
...
+
+    _build_attribute_vector(raw_dictionary, raw_values);
+  }
+
+  void _build_attribute_vector(std::vector<T>& raw_dictionary, std::vector<T>& all_values) {
+    std::vector<uint32_t> raw_vector{};
+    raw_vector.reserve(all_values.size());
+    size_t size = all_values.size();
+
+    const int required_width = std::ceil(std::log2(size));
```
Comment (Bensk1) [inline src/lib/storage/dictionary_segment.hpp]: This is harder to understand that it should be. Apparently, in the end, `size` refers to the size of the original attribute vector while the width should be calculated based on the size of the dictionary, i.e., the unique values.

File: src/lib/storage/dictionary_segment.hpp [hunk 12]
```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
...
+
+    if (required_width <= 8) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint8_t>>(size);
```
Comment (Bensk1) [inline src/lib/storage/dictionary_segment.hpp]: If `size` is `256` you would choose an 8bit AttributeVector which would leave no space for `INVALID_VALUE_ID`.

File: src/lib/storage/dictionary_segment.hpp [hunk 13]
```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
...
+  size_t unique_values_count() const { return _dictionary->size(); }
 
   // return the number of entries
-  ChunkOffset size() const override;
+  ChunkOffset size() const { return _attribute_vector->size(); }
 
   // returns the calculated memory usage
-  size_t estimate_memory_usage() const final;
+  size_t estimate_memory_usage() const final {
+    auto attributeVecMem = _attribute_vector->width() * _attribute_vector->size();
```
Comment (mweisgut) [inline src/lib/storage/dictionary_segment.hpp:103]: Please follow the contribution guidelines: `Variables: lowercase separated by underscores, e.g., chunk_size.`

File: src/lib/storage/fixed_size_attribute_vector.hpp [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+  // we overwrite the copy constructor
+  FixedSizeAttributeVector(FixedSizeAttributeVector&&) = default;
+  FixedSizeAttributeVector& operator=(FixedSizeAttributeVector&&) = default;
+
+  // returns the value id at a given position
+  ValueID get(const size_t i) const { return static_cast<ValueID>(_attributes.at(i)); }
+
+  // sets the value id at a given position
+  void set(const size_t index, const ValueID value_id) {
+    // [] operator does somehow not change the size of the vector
```
Comment (Bensk1) [inline src/lib/storage/fixed_size_attribute_vector.hpp]: It's not supposed to change the size.

File: src/lib/storage/fixed_size_attribute_vector.hpp [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+  ~FixedSizeAttributeVector() = default;
+
+  // we need to explicitly set the move constructor to default when
+  // we overwrite the copy constructor
+  FixedSizeAttributeVector(FixedSizeAttributeVector&&) = default;
+  FixedSizeAttributeVector& operator=(FixedSizeAttributeVector&&) = default;
+
+  // returns the value id at a given position
+  ValueID get(const size_t i) const { return static_cast<ValueID>(_attributes.at(i)); }
```
Comment (mweisgut) [inline src/lib/storage/fixed_size_attribute_vector.hpp:25]: Choose clear and concise names, and avoid, e.g., i, j, ch_ptr.

File: src/lib/storage/fixed_size_attribute_vector.hpp [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+  // sets the value id at a given position
+  void set(const size_t index, const ValueID value_id) {
+    // [] operator does somehow not change the size of the vector
+    // according to this stackoverflow answer it's considered bad practice to just use the [] operator
+    // https://stackoverflow.com/questions/31372809/c-vector-size-is-returning-zero
+    // nevertheless, this does not feel right.
```
[the remark this answers is not in the document]
Comment (Bensk1) [reply src/lib/storage/fixed_size_attribute_vector.hpp]: Maybe, you have mixed up `reserve` and `resize`.

File: src/lib/storage/table.cpp [hunk 17]
```diff
@@ -81,23 +83,71 @@ const std::string& Table::column_name(const ColumnID column_id) const { return _
...
+    const auto compressed_segment = std::make_shared<DictionarySegment<ColumnDataType>>(column_segment);
+    compressed_chunk->add_segment(compressed_segment);
```
Comment (Bensk1) [inline src/lib/storage/table.cpp:149]: `add_segment` could theoretically be called from multiple concurrent threads at the same time. See Moodle for a [similar](https://moodle.hpi.de/mod/hsuforum/discuss.php?d=43) discussion.

File: src/lib/storage/table.cpp [hunk 18]
```diff
@@ -81,23 +83,71 @@ const std::string& Table::column_name(const ColumnID column_id) const { return _
...
+  std::lock_guard<std::mutex> lock(_chunk_lock);
+  auto& uncompressed_chunk = _get_chunk(chunk_id);
+  Assert(uncompressed_chunk.size() == target_chunk_size(),
+         "Attempt to compress chunk that is not yet completely filled.");
+
+  std::shared_ptr<Chunk> compressed_chunk = _compress_multithreaded(uncompressed_chunk);
+  _chunks[chunk_id] = compressed_chunk;
+}
+
+std::shared_ptr<Chunk> Table::_compress_multithreaded(Chunk& uncompressed_chunk) {
```
Comment (mweisgut) [inline src/lib/storage/table.cpp:121]: `const Chunk& uncompressed_chunk`?
